### PR TITLE
Question Authoring PR Draft

### LINKED ISSUE #343: Add "Question Authoring" Boards
### Overview
On board creation, allow teachers to select either "brainstorm" or "question authoring" boards. If question authoring board is selected, all post elements are preserved (e.g., title, comments, tags, upvotes, but the message is now either an open response question prompt or a multiple choice editor. A selector in the Add Post editor can choose whether the post is an open response prompt or a multiple choice question. Both are saved as JSON objects in SCORE authoring format. 

### Toggle “Brainstorm/Question Authoring” Board Type

**Figure 1. Add New Board Configurations**
Screen Shot 2022-08-28 at 3 23 39 PM 

- In new board configurations, create the heading: Board Type
- Under the board type heading add toggle buttons: “Brainstorm” and “Question Authoring”

### Display Multiple Choice UI by Default for New Post Editor

When a new post is being created in a question authoring board, create a toggle button between "multiple choice" and "open response" 

Screen Shot 2022-08-28 at 3 32 27 PM

All the posts are either:

- Multiple Choice or Open Response
- When users select the question type, the UI will be updated to match selection
- Multiple Choice can be the default option 
- (We can port over SCORE’s multiple choice editor UI, advanced features such as image uploads, etc. can be removed)
- The Multiple Choice question prompt and choices will be saved as JSON as per the SCORE format 
- In the future, the JSON will be imported to SCORE as a question

### Display Open Response Posts when Enabled for New Post Editor

Screen Shot 2022-08-28 at 3 35 06 PM
When open response is selected, we may use the classic message box, however:

- The prompt instead of saying “Message” will say something else, such as “Open response question prompt”
- The authored open response prompt will be saved as JSON as per the SCORE format
- Again, in the future, the JSON will be imported to SCORE as a question

### Next step - Add MathJax/ASCIIMath Previewer for Multiple Choice and Open Response
Screen Shot 2022-08-28 at 3 57 42 PM
---
Screen Shot 2022-08-28 at 4 00 30 PM

### PR DESCRIPTION
**Pull request by LunarFang416**
## Details

- Implemented Question Authoring functionality on top of teh existing post creation feature
- Open response type posts are the same as before (but with different message prompt)
- Users can now create multiple choice questions in board, with a minimum of 2 options upto 6 options with only 1 correct answer
- Editing Multiple choice posts is also possible. This utilizes the existing `add-post-modal`. Options can be edited by the teacher to decrease/increase the number of options, or change question title. 

Tests:
- Could not do much stress testing regarding socket updates, however changes to the backend were minute, so chances that breaks occur are very slim
- However, I still made sure to test basic operations like, creating post, deleting post, adding tags, removing tags, which were successful.

Closes #343

### CODE REVIEW

**Review by JoelWiebe — CHANGES_REQUESTED** [review]
This is sooo cool! I really like the work you did with "Check Answer" for a created post. That's a really neat way for students to experience taking the questions.

Suggestions:
 1. I'm sorry, I think I gave some unclear instructions. At board creation, where we have the two types of board, could we change "Multiple Choice" and "Open Response" to "Brainstorm board" and "Question Authoring"? 
 * Brainstorm would be the classic board with "title" and "message" fields in its posts
 * Question Authoring boards would have the option to create either Multiple Choice questions or open response questions when adding new posts to the board. 
Screen Shot 2022-09-15 at 2 56 23 PM

 2. Can we choose between post types when creating a new post in a Question Authoring board? We could do this perhaps either of two ways. Maybe this could work something like the following: when the add post modal pops up, there is an option above the title for Multiple Choice or Open Response, clicking on the Multiple Choice shows (1) the title, (2) multiple choice question prompt (NEW to your design), and (3) multiple choice creation tool; clicking Open Response option essentially dynamically (1) hides the multiple choice tool and (2) changes the question prompt to the open response question prompt (as shown below). Multiple choice could be the default option.
Screen Shot 2022-09-15 at 2 53 39 PM

Screen Shot 2022-09-15 at 2 54 30 PM

 3. I just noticed that the manage group icon is duplicated in the project dashboard -- probably from a merge at some point
Screen Shot 2022-09-15 at 3 04 24 PM

Really great work! Sorry about the confusion. Is something like the above possible?

**Comment by LunarFang416** [comment]
@JoelWiebe Made the changes that you requested. Hopefully it works smoothly now. 👍

File: frontend/src/app/components/post-modal/post-modal.component.ts [hunk 1]
```diff
@@ -306,6 +328,58 @@ export class PostModalComponent {
     this.expandedUpvotesView = false;
   }
 
+  answerMultipleChoice() {
+    this.submitMultipleChoiceAnswer = true;
+    if (this.isMultipleChoiceSelected && this.selectedMultipleChoice) {
+      if (this.selectedMultipleChoice.correct) {
+        this.snackbarService.queueSnackbar('Correct Answer!');
+      } else {
+        this.snackbarService.queueSnackbar('Incorrect, Try Again.');
```
Comment (JoelWiebe) [inline frontend/src/app/components/post-modal/post-modal.component.ts:337]: Maybe either "Incorrect. Try again." (with a period after "Incorrect") or "Incorrect, please try again" (keeping the comma after "Incorrect"

**Comment by JoelWiebe** [comment]
@LunarFang416 This is simply amazing! Well done.

Really looking forward to trying this out in a classroom. I may pitch this to a teacher in a graduate level class too. 

1. Is there a way we could let students know the requirements for being able to submit a multiple choice question? For instance, could they receive an error if they (1) do not have enough multiple choice options, (2) forgot to select a correct answer, or (3) did not create a prompt? 

2. Also, could we make the question field required for question authoring board open response question prompts?

These are lower priority, so I'm happy to come back to these later if you don't have time to work on these right now.

**Comment by JoelWiebe** [comment]
Oh yeah, I guess we can remove the console logs too. Thanks for adding those, they are very helpful for testing.

**Comment by LunarFang416** [comment]
@JoelWiebe Just added the validation for the inputs when creating a post in the question authoring boards.

**Review by JoelWiebe — APPROVED** [review]
Great work!

**Comment by JoelWiebe** [comment]
@DCH25 @Ammar-T Would one of you have a quick chance to review this?

File: backend/src/models/Post.ts [hunk 2]
```diff
@@ -49,12 +65,18 @@ export class PostModel {
   @prop({ enum: PostType, type: String, required: true })
   public type!: PostType;
 
+  @prop({ enum: PostCreationType, type: String, required: true })
+  public postCreationType!: PostCreationType;
```
Comment (Ammar-T) [inline backend/src/models/Post.ts:69]: I think `contentType` would be a better name for this field.

**Review by Ammar-T — APPROVED** [review]
Just the small field renaming and this looks fine to merge.

**Pull request merged**